atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientImpl.java

```diff
@@ -107,18 +175,56 @@ public CASResult cas(TableReference tableReference,
             ConsistencyLevel commit_consistency_level)
             throws InvalidRequestException, UnavailableException, TimedOutException, TException {
         String internalTableName = AbstractKeyValueService.internalTableName(tableReference);
-        return client.cas(key, internalTableName, expected, updates, serial_consistency_level,
-                commit_consistency_level);
+
+        try (CloseableTrace trace = startLocalTrace("client.cas(table {})",
+                LoggingArgs.safeTableOrPlaceholder(tableReference))) {
+            return KvsProfilingLogger.maybeLog(
+                    (KvsProfilingLogger.CallableCheckedException<CASResult, TException>)
+                            () -> client.cas(key, internalTableName, expected, updates, serial_consistency_level,
+                                    commit_consistency_level),
+                    (logger, timer) -> logger.log("client.cas({}) took {} ms",
+                            LoggingArgs.tableRef(tableReference),
+                            LoggingArgs.durationMillis(timer)));
+        }
     }
 
     @Override
-    public CqlResult execute_cql3_query(ByteBuffer query, Compression compression, ConsistencyLevel consistency)
+    public CqlResult execute_cql3_query(CqlQuery cqlQuery,
+            Compression compression,
+            ConsistencyLevel consistency)
             throws InvalidRequestException, UnavailableException, TimedOutException, SchemaDisagreementException,
             TException {
-        return client.execute_cql3_query(query, compression, consistency);
+        ByteBuffer queryBytes = ByteBuffer.wrap(cqlQuery.fullQuery().getBytes(StandardCharsets.UTF_8));
+
+        try (CloseableTrace trace = startLocalTrace("cqlExecutor.execute_cql3_query(query {})", cqlQuery)) {
+            return KvsProfilingLogger.maybeLog(
+                    (KvsProfilingLogger.CallableCheckedException<CqlResult, TException>)
+                            () -> client.execute_cql3_query(queryBytes, compression, consistency),
+                    (logger, timer) -> this.logSlowResult(cqlQuery, logger, timer),
+                    this::logResultSize);
+        }
+    }
+
+    private void logSlowResult(CqlQuery cqlQuery, KvsProfilingLogger.LoggingFunction log, Stopwatch timer) {
```

Comment:
would prefer to keep this logic inside `CqlQuery` (we can just have a `logTo(LoggingFunction, Stopwatch)`, method)